src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,598 @@
+      instr_it->type=ASSIGN;
+      instr_it->code=assignment;
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::instrument_function_calls
+
+Inputs:
+
+Outputs:
+
+Purpose: after each function call g() in function f 
+         adds f#exception_value=g#exception_value;
+
+\*******************************************************************/
+
+void remove_exceptionst::instrument_function_calls(
+  const goto_functionst::function_mapt::iterator &func_it)
+{
+  const irep_idt &caller_id=func_it->first;
+  goto_programt &goto_program=func_it->second.body;
+
+  if(goto_program.empty())
+    return;
+
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    if(instr_it->is_function_call())
+    {
+      code_function_callt &function_call=to_code_function_call(instr_it->code);
+      const irep_idt &callee_id=
+        to_symbol_expr(function_call.function()).get_identifier();
+
+      // can exceptions escape?
+      if(symbol_table.has_symbol(id2string(callee_id)+EXC_SUFFIX) &&
+        symbol_table.has_symbol(id2string(caller_id)+EXC_SUFFIX))
+      {
+        const symbolt &callee=
+          symbol_table.lookup(id2string(callee_id)+EXC_SUFFIX);
+        const symbolt &caller=
+          symbol_table.lookup(id2string(caller_id)+EXC_SUFFIX);
+
+        symbol_exprt rhs_expr;
+        rhs_expr.set_identifier(id2string(callee_id)+EXC_SUFFIX);
+        rhs_expr.type()=callee.type;
+
+        symbol_exprt lhs_expr;
+        lhs_expr.set_identifier(id2string(caller_id)+EXC_SUFFIX);
+        lhs_expr.type()=caller.type;
+
+        goto_programt::targett t=goto_program.insert_after(instr_it);
+        t->make_assignment();
+        t->source_location=instr_it->source_location;
+        t->code=code_assignt(lhs_expr, rhs_expr);
+        t->function=instr_it->function;
+      }
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::instrument_exception_handlers
+
+Inputs:
+
+Outputs:
+
+Purpose: at the beginning of each handler in function f 
+         adds exc=f#exception_value; f#exception_value=NULL;
+
+\*******************************************************************/
+
+void remove_exceptionst::instrument_exception_handlers(
+  const goto_functionst::function_mapt::iterator &func_it)
+{
+  const irep_idt &function_id=func_it->first;
+  goto_programt &goto_program=func_it->second.body;
+
+  if(goto_program.empty())
+    return;
+
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    // is this a handler
+    if(instr_it->type==CATCH && instr_it->code.has_operands())
+    {
+      // retrieve the exception variable
+      const irept &exception=instr_it->code.op0();
+
+      if(symbol_table.has_symbol(id2string(function_id)+EXC_SUFFIX))
+      {
+        const symbolt &function_symbol=
+          symbol_table.lookup(id2string(function_id)+EXC_SUFFIX);
```

Comment:
Just a thought: You can do `has_symbol` and `lookup` in a single step by using the 2-argument version of `lookup` that takes a reference-to-a-pointer as second argument.